MSI source Import

larsroettig opened the pull request:
### Description
Import module for Multi Source Inventory

### Manual testing scenarios
1. Create a CSV file with e.g. following content
```
"source","sku","qty","status"
1,"test",60,1
2,"test",2,0
3,"test",34,1
1,"foobar",55,1
2,"foobar",44,1
```
2. Go to `System > Import` page in Magento backend
3. Choose Entity Type `Stock Sources` and import the CSV with append behavior

### TODOs
- [x] Implement Import Entity for Stock Sources
   - [x] File validation
   - [x] DELETE behavior
   - [x] REPLACE behavior
   - [x] APPEND behavior
- [x] Integration tests for Import

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

vadimjustus committed:
MSI: Initial implementation of Magento_InventoryImportExport module

vadimjustus committed:
MSI: Implement append bahaviour import for source items

larsroettig committed:
Merge pull request #115 from vadimjustus/msi-source-import-export

WIP: MSI source import/export

Commit:
[task] #115 Refactoring
- Add Documentation
- phpcs fixes
- Impelement phpunit test for ValidatorChain

vadimjustus committed:
MSI: Change method accessability to private

vadimjustus committed:
MSI: Implement intergration test for import entity

Commit:
wip

Commit:
Merge remote-tracking branch 'origin/develop' into msi-source-import-export

Commit:
wip

Commit:
MSI source: Impelment DelteMultiple Model and Tests

Commit:
Remove not required fixture from methodes

Commit:
MSI source: Bugfix for Indexer Integration test

Commit:
Merge remote-tracking branch 'origin/develop' into msi-source-import-export

Commit:
MSI source:
- Add SPI for Delete Multiple
- Improve Delte Comand
- Implement SPI

Commit:
MSI source:
- Bugfix for Intigrationstest to  new source_ids
- Impelment new test case for delete behavior

Commit:
MSI source:
- Refactoring: Add Helperfunction for convert
- Impelment Replace Comand
- Impelment Replace Comand Test

TODO:
- Implement clenup for Replace Comand

Commit:
phpmd and phpcs fixes

Commit:
MSI: Add Testcase for testImportDataWithWrongBehavior
- phpcs fixes
- phpmd fixes

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/DeleteMultiple.php:
```diff
@@ -0,0 +1,72 @@
+ * Implementation of SourceItem delete multiple operation for specific db layer
+ * Delete Multiple used here for performance efficient purposes over single delete operation
+ */
+class DeleteMultiple
```
Looks like we need to remove
`SourceItemRepositoryInterface::delete(SourceItemInterface $sourceItem); `
method, as we have 
`SourceItemsDeleteInterface::execute(array $sourceItems);`

which do the same

larsroettig replied:
Hi @maghamed if we remove we will break the current delete function
behaviour. Because the current **Delete Multiple** removes all entries from the Database by only matching SKU. 


If you mean it is okay I open for this an extra PR:
https://github.com/magento-engcom/msi/pull/140

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/SourceItem/DeleteMultiple.php:
```diff
@@ -0,0 +1,72 @@
+     * @param SourceItemInterface[] $sourceItems
+     * @return int[]
+     */
+    private function getSourceIds($sourceItems): array
```
name of the method is not correct, as it returns source_id and sku list assigned to it

maghamed commented on app/code/Magento/Inventory/Model/SourceItem/Command/SourceItemsDelete.php:
```diff
@@ -0,0 +1,57 @@
+    {
+        if (empty($sourceItems)) {
+            throw new InputException(__('Input data is empty'));
+        }
```
not sure that we need to throw an exception when someone tries to delete empty array of source Items. 

For example, you have a replace strategy and you want to delete existing SourceItems and substitute with new ones, but on clean Magento installation with empty DB - delete would have an empty set

maghamed commented on app/code/Magento/Inventory/Test/Integration/Model/ResourceModel/SourceItem/DeleteMultipleTest.php:
```diff
@@ -0,0 +1,80 @@
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Inventory\Model\ResourceModel\SourceItem;
```
all the tests should be stored under the namespace 

`Magento\Inventory\Test\Unit|Integration\Model`

not mixed with code. Such tests would not be even run by Travis

maghamed commented on app/code/Magento/InventoryApi/Api/SourceItemsDeleteInterface.php:
```diff
@@ -0,0 +1,28 @@
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\InventoryApi\Api;
```
don't forget to use strict typing for newly created code

maghamed commented on app/code/Magento/InventoryImportExport/Helper/SourceItemConvert.php:
```diff
@@ -0,0 +1,54 @@
+use Magento\InventoryApi\Api\Data\SourceItemInterfaceFactory;
+use Magento\InventoryImportExport\Model\Import\Sources;
+
+class SourceItemConvert
```
no need to Introduce Helper in the new Code.

Helpers should be discouraged in newly created code.
You could create SourceItemConvert under the Model namespace instead

maghamed commented on app/code/Magento/InventoryImportExport/Model/Import/Command/Append.php:
```diff
@@ -0,0 +1,46 @@
+    private $sourceItemsSave;
+
+    /**
+     * @param SourceItemBuilder
```
wrong params in annotation , don't match to real params in constructor

maghamed commented on app/code/Magento/InventoryImportExport/Model/Import/Command/CommandException.php:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Exception for command execution fails
+ */
+class CommandException extends LocalizedException
```
If we introduce new Exception its name should be expressive enough. 

Currently `CommandException ` does not say a lot

maghamed commented on app/code/Magento/InventoryImportExport/Model/Import/Command/CommandInterface.php:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @api
+ */
+interface CommandInterface
```
if we introduce either API or SPI we should mark them with @api

larsroettig replied:
@api already implemented

maghamed commented on app/code/Magento/InventoryImportExport/Model/Import/Serializer/LegacyJsonHelperInterface.php:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @api
+ */
+interface LegacyJsonHelperInterface
```
why do we need to introduce this interface? especially if it's already legacy? 

If you want to make Backward Compatible change, legacy class 
`Magento\Framework\Json\Helper\Data` should be adapted as well.

Because currently, you are breaking the contract, as AbstractEntity accept 
 `      \Magento\Framework\Json\Helper\Data $jsonHelper,`
But you pass 
` SerializerInterface $jsonHelper, ` instead

maghamed commented on app/code/Magento/InventoryImportExport/Model/Import/Serializer/Json.php:
```diff
@@ -0,0 +1,78 @@
+     */
+    public function jsonEncode($valueToEncode)
+    {
+        return $this->serializer->serialize($valueToEncode);
```
this method should be implemented proxying call to 
`$this->serialize($valueToEncode)`

maghamed commented on app/code/Magento/InventoryImportExport/Model/Import/Serializer/Json.php:
```diff
@@ -0,0 +1,78 @@
+     */
+    public function jsonDecode($encodedValue)
+    {
+        return $this->serializer->unserialize($encodedValue);
```
this method should be implemented proxying call to 
`$this->unserialize($valueToEncode)`

to make all Pluginization of SerializerInterface work correctly

maghamed commented on app/code/Magento/InventoryImportExport/Model/Import/Validator/SkuValidator.php:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Extension point for row validation
+ *
+ * @api
```
why do we have @api on class implementation?

maghamed commented on app/code/Magento/InventoryImportExport/Model/Import/Validator/SourceValidator.php:
```diff
@@ -0,0 +1,88 @@
+/**
+ * Extension point for source validation
+ *
+ * @api
```
no need to use @api here

Commit:
MSI:
- declare strict_types fot Import/Export Module
- Add Replace Comand
- Improve Coverrage
- phpmd und phpcs

Commit:
MSI:
- declare strict_types fot Import/Export Module
- phpmd and phpunit fix
- review fixes

vadimjustus committed:
MSI: Remove LegacyJsonHelperInterface

Commit:
MSI: #131
- Add Export Command

vadimjustus committed:
Merge branch 'msi-source-import-export' into msi-source-export

Commit:
wip

Commit:
Merge branch 'msi-source-export' of github.com:magento-engcom/magento2 into msi-source-export

Commit:
WIP bugfix for export filter

Commit:
wip

vadimjustus committed:
MIS: Implement filtering for export functionality

Commit:
add filter type

vadimjustus committed:
Merge branch 'msi-source-export' of github.com:magento-engcom/msi into msi-source-export

vadimjustus committed:
MSI: Extend filtering logic for stock_source export

vadimjustus committed:
MSI: remove useless code

vadimjustus committed:
MSI: Refactore filtering logic in order to reduce complexity

vadimjustus committed:
MSI: Fix intergration tests

vadimjustus committed:
MSI: Implement integration tests for stock_source export entity

vadimjustus committed:
MSI: Refactore getAttributeCollection method for stock_sources export model

vadimjustus committed:
MSI: Refactore filtering logic for stock_source export entity

vadimjustus committed:
MSI: Static improvements

vadimjustus committed:
MSI: Refactore stock_source export entity dependencies

vadimjustus committed:
MSI: Fix static code style issues

vadimjustus committed:
MSI: Move SPI interfaces for export to model directory

vadimjustus committed:
MSI: remove useless datetime filter processor

maghamed committed:
MSI Import/Export: added comments for SPI interfaces

maghamed committed:
Merge pull request #147 from magento-engcom/msi-source-export

Msi source export
